usb/usb_host_hid/hid_host.c

```diff
@@ -512,51 +596,53 @@ static esp_err_t hid_host_interface_shutdown(hid_host_device_handle_t hid_dev_ha
 
     HID_RETURN_ON_INVALID_ARG(iface);
 
-    iface->state = HID_INTERFACE_STATE_WAIT_USER_DELETION;
-
     // Not force shutdown and Interface was started by user
     if (!force && iface->user_cb) {
+        iface->state = HID_INTERFACE_STATE_WAIT_USER_DELETION;
         hid_host_user_interface_callback(iface, HID_HOST_INTERFACE_EVENT_DISCONNECTED);
-        return ESP_OK;
     }
 
-    // Interface has not been opened, delete it right away
-    return hid_host_device_close(iface);
+    return ESP_OK;
 }
 
 /**
- * @brief Destroy a list of available interfaces in RAM
+ * @brief Destroy a list of available interfaces and devices in RAM
  *
  * @return esp_err_t
  */
-static esp_err_t hid_host_interface_list_destroy(void)
+static esp_err_t hid_host_lists_destroy(void)
 {
-    bool active_shutdown = false;
+    hid_device_t *device = NULL;
+    hid_iface_t *interface = NULL;
 
-    hid_iface_t *iface = STAILQ_FIRST(&s_hid_driver->hid_ifaces_tailq);
-    hid_iface_t *tmp = NULL;
+    // Interface list
+    HID_ENTER_CRITICAL();
+    while (!STAILQ_EMPTY(&s_hid_driver->hid_ifaces_tailq)) {
+        interface = STAILQ_FIRST(&s_hid_driver->hid_ifaces_tailq);
+        HID_EXIT_CRITICAL();
 
-    while (iface != NULL) {
-        tmp = STAILQ_NEXT(iface, tailq_entry);
-        // free interface
-        if (HID_INTERFACE_STATE_NOT_INITIALIZED != iface->state) {
-            active_shutdown = true;
-            ESP_ERROR_CHECK( hid_host_device_close(iface) );
-            ESP_ERROR_CHECK( hid_host_interface_shutdown(iface, true));
+        if (interface->state > HID_INTERFACE_STATE_IDLE) {
+            ESP_ERROR_CHECK( hid_host_device_close(interface) );
         }
 
+        ESP_ERROR_CHECK( hid_host_interface_shutdown(interface, true));
+
         HID_ENTER_CRITICAL();
-        iface->state = HID_INTERFACE_STATE_NOT_INITIALIZED;
-        STAILQ_REMOVE(&s_hid_driver->hid_ifaces_tailq, iface, hid_interface, tailq_entry);
-        free(iface);
+        _hid_host_remove_interface(interface);
```

Comment:
There's a gap in the critical section here. Thus, during that gap, another thread could already have removed the interface from the `hid_ifaces_tailq`. What if we removed it from the list first and then carried out the shut down.

Reply:
Yes, I know that. And still I didn't come up with any better idea how to fix it. 
There is a logical problem with architecture, that is why I decided to change it some time ago. 
I explained everything more detailed in the document, which I attached to the PR for next version: https://github.com/espressif/idf-extra-components/pull/214#issuecomment-1643770574

The whole idea to keep list of usb device opened device is strange. 
That is why there is a two-way choice during the hid_host_unsinstall(). 

But I'll try to figure something out to fix this version. 
Thanks.

Reply:
Change the way how list is freed. 
Yes, there is still a gap in critical section, but in case if the some elements will be absent, should be no problem. 

I guess this is the best option to do in case of current logic.